nrf5340_audio: enable passive scanning to find name in scan response

mringwal opened the pull request:
I'm trying to connect to OnePlus Buds Pro 2 and finally figured out why the unicast demo didn't connect to it: the device name is not in the regular advertising, instead it's in the scan response . This PR enables passive scanning.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/nrfconnect/sdk-nrf?pullRequest=10717) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/nrfconnect/sdk-nrf?pullRequest=10717) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/nrfconnect/sdk-nrf?pullRequest=10717) it.</sub>

NordicBuilder commented:
Thank you for your contribution!
It seems you are not a member of the nrfconnect GitHub organization. External contributions are handled as follows:
Large contributions, affecting multiple subsystems for example, may be rejected if they are complex, may introduce regressions due to lack of test coverage, or if they are not consistent with the architecture of nRF Connect SDK.
PRs will be run in our continuous integration (CI) test system.
If CI passes, PRs will be tagged for review and merged on successful completion of review. You may be asked to make some modifications to your contribution during review.
If CI fails, PRs may be rejected or may be tagged for review and rework.
PRs that become outdated due to other changes in the repository may be rejected or rework requested.
External contributions will be prioritized for review based on the relevance to current development efforts in nRF Connect SDK.  Bug fix PRs will be prioritized.
You may raise issues or ask for help from our Technical Support team by visiting https://devzone.nordicsemi.com/.

*Note: This comment is automatically posted and updated by the Contribs GitHub Action.*

alexsven commented:
Hey @mringwal, thank you for the PR. Looking at the changes it doesn't really match the commit msg; you are enabling active scan, not passive. We can't hard code that into the project, but it's a good point that we should be able to configure different scan types. 

I suggest adding a Kconfig for this.

mringwal commented:
HI @alexsven Sorry for the line re-ordering. Fixed in the latest force-push. 
I've also updated the PR title to reflect the changes.

If active scan is enabled, it will find the device name in both advertising data as well as scan data. 

Do you see a problem in using active scan in general?
If yes, a Kconfig flag could be used as suggested by you.

alexsven commented:
> HI @alexsven Sorry for the line re-ordering. Fixed in the latest force-push. I've also updated the PR title to reflect the changes.
> 
> If active scan is enabled, it will find the device name in both advertising data as well as scan data.
> 
> Do you see a problem in using active scan in general? If yes, a Kconfig flag could be used as suggested by you.

It's using a lot more power than the passive  scan. So if the active scan is not needed we should use passive scanning

alexsven commented on applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:
```diff
@@ -1009,8 +1009,8 @@ static void on_device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
 		return;
-	} else if ((type == BT_GAP_ADV_TYPE_ADV_IND || type == BT_GAP_ADV_TYPE_EXT_ADV) &&
-		   (bonded_num < CONFIG_BT_MAX_PAIRED)) {
+	} else if ((type == BT_GAP_ADV_TYPE_ADV_IND || type == BT_GAP_ADV_TYPE_EXT_ADV || (type == BT_GAP_ADV_TYPE_SCAN_RSP)) &&
+	   		(bonded_num < CONFIG_BT_MAX_PAIRED)) {
```
Either add parenthesis to all of the 'type == BT_GAP...' or remove the ones for your addition, so that it looks more consistent. There is also the formatting, we use clang for out files, and the line shouldn't be more than 100 characters. I've posted the correct formatting below
```suggestion
	} else if ((type == BT_GAP_ADV_TYPE_ADV_IND || type == BT_GAP_ADV_TYPE_EXT_ADV ||
		    type == BT_GAP_ADV_TYPE_SCAN_RSP) &&
		   (bonded_num < CONFIG_BT_MAX_PAIRED)) {
```

mringwal commented:
Ok. As zephyr / west / Kconfig beginner, I've added a Kconfig option to enable active scan and then used it to select active scan if requested.

alexsven commented:
> Ok. As zephyr / west / Kconfig beginner, I've added a Kconfig option to enable active scan and then used it to select active scan if requested.

Very nice! Two final comments is that the Kconfig choice you added should be in the bluetooth Kconfig. So in nrf/applications/nrf5340_audio/src/bluetooth/Kconfig. Also, the new define from le_audio.h should be applied to le_audio_bis_headset.c as well. 

Other than that this looks great, I like the switch you added to the adv_parse.

mringwal commented:
Thanks for the review. I've moved the kconfig flag to bluetooth. For broadcast, different flags/macros are used so I had to introduce another one. 

Btw. the broadcast gateway doesn't build for me as `CONFIG_BT_AUDIO_BROADCAST_SRC_SUBGROUP_COUNT` is not defined, which doesn't seem to be related to my changes that enable active scanning.

koffes commented on applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:
```diff
@@ -1003,16 +1003,23 @@ static void ad_parse(struct net_buf_simple *p_ad, const bt_addr_le_t *addr)
 {
-	/* Direct advertising has no payload, so no need to parse */
-	if (type == BT_GAP_ADV_TYPE_ADV_DIRECT_IND) {
+	switch(type){
```
Nice change. This is easier to read.

mringwal replied:
Thanks

koffes commented on applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:
```diff
@@ -1003,16 +1003,23 @@ static void ad_parse(struct net_buf_simple *p_ad, const bt_addr_le_t *addr)
-	} else if ((type == BT_GAP_ADV_TYPE_ADV_IND || type == BT_GAP_ADV_TYPE_EXT_ADV) &&
-		   (bonded_num < CONFIG_BT_MAX_PAIRED)) {
+		break;
+	case BT_GAP_ADV_TYPE_ADV_IND:
```
The code standard requires that /* Fall through */ is added to these cases to underline that this is intentional.

koffes commented on applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c:
```diff
@@ -1003,16 +1003,23 @@ static void ad_parse(struct net_buf_simple *p_ad, const bt_addr_le_t *addr)
+		break;
+	case BT_GAP_ADV_TYPE_ADV_IND:
+	case BT_GAP_ADV_TYPE_EXT_ADV:
+	case BT_GAP_ADV_TYPE_SCAN_RSP:
```
This SCAN_RSP type did not previously end up in this if. Can you comment?

mringwal replied:
I'll move the SCAN_RSP addition to a later commit then.

koffes replied:
The SCAN_RSP

mringwal replied:
The SCAN_RSP was actually done in a later commit. After the squash, this detail was gone. :)

koffes commented on applications/nrf5340_audio/src/bluetooth/le_audio.h:
```diff
@@ -16,6 +16,14 @@
 #define LE_AUDIO_SDU_SIZE_OCTETS(bitrate) (bitrate / (1000000 / CONFIG_AUDIO_FRAME_DURATION_US) / 8)
 
+#if (CONFIG_SCAN_MODE_ACTIVE)
+#define BT_LE_SCAN_TYPE    BT_LE_SCAN_TYPE_ACTIVE
```
Since these defines are set here, we don't recommend using BT_LE_* as these are defined elsewhere.

mringwal replied:
OK, I'll use a project specific prefix then.

koffes commented on applications/nrf5340_audio/src/bluetooth/le_audio.h:
```diff
@@ -16,6 +16,14 @@
+#if (CONFIG_SCAN_MODE_ACTIVE)
+#define BT_LE_SCAN_TYPE    BT_LE_SCAN_TYPE_ACTIVE
+#define BT_LE_SCAN_PARAMS  BT_LE_SCAN_ACTIVE
+#else
```
Some nitpick, but these changes make the code more resilient from outside/upstream changes.
```suggestion
#elif (CONFIG_SCAN_MODE_PASSIVE)
```
And then an #error .

mringwal replied:
Can add that, too.

alexsven commented:
> Thanks for the review. I've moved the kconfig flag to bluetooth. For broadcast, different flags/macros are used so I had to introduce another one.
> 
> Btw. the broadcast gateway doesn't build for me as `CONFIG_BT_AUDIO_BROADCAST_SRC_SUBGROUP_COUNT` is not defined, which doesn't seem to be related to my changes that enable active scanning.

That might be a case of mismatched versions of zephyr. Using 'west update', as long as you haven't changed west.yml should fix that. I know that they have renamed a lot of the CONFIGs in zephyr upstream lately.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-10717
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

mringwal commented:
I hope I caught all requested changes.

rick1082 commented on applications/nrf5340_audio/src/audio/streamctrl.c:
```diff
@@ -365,6 +365,7 @@ static void le_audio_evt_handler(enum le_audio_evt_type event)
 	int ret;
 
 	LOG_DBG("Received event = %d, current state = %d", event, strm_state);
+	uint32_t pres_delay;
```
Would be great to move line 368 to line 366 and leave a blank after that.

mringwal replied:
Hi @rick1082 - I've made no changes to that. I only wanted to add an option to enable active scanning. 

Is it easier for you if I make this change in this PR now?

mringwal replied:
Done. I forgot that I made the changes to get it to compile. Yes, looks better that way.

alexsven commented on applications/nrf5340_audio/src/bluetooth/le_audio_bis_headset.c:
```diff
@@ -633,7 +633,7 @@ int le_audio_enable(le_audio_receive_cb recv_cb, le_audio_timestamp_cb timestmp_
 	LOG_INF("Scanning for broadcast sources");
 
-	ret = bt_audio_broadcast_sink_scan_start(BT_LE_SCAN_PASSIVE);
+	ret = bt_audio_broadcast_sink_scan_start(BT_LE_SCAN_PARAMS);
```
```suggestion
	ret = bt_audio_broadcast_sink_scan_start(NRF5340_AUDIO_GATEWAY_SCAN_PARAMS);
```

alexsven replied:
It was failing compile because of this

mringwal replied:
Thanks. Missed that one. Fixed in the last force-push/squash action. Let's see if there's still compliance issues.

alexsven commented:
Commits should be squashed to a single commit and to pass compliance it needs to follow the guidelines outlined here: https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines

You can also click the Compliance Details to see the things it is complaining about. 

TLDR: Squash into one commit, header should be "applications: nrf5340_audio: Add Kconfig for scan type". There are also some whitespaces at end of line in some of the files.

alexsven commented:
> The Compliance test doesn't start because of "Waiting for status to be reported". No idea, clicking 're-request review'. Other options?

I think it's because I have to approve that the CI will run, and it has to be done every time there is a new commit. Approved that it will be run now.

alexsven commented:
@mringwal: I've asked to get it merged now, thank you for your contribution.
